plutus-ledger/src/Ledger/Fee.hs

```diff
@@ -1,38 +1,78 @@
-{-# LANGUAGE DeriveAnyClass #-}
-{-# LANGUAGE NamedFieldPuns #-}
-
--- | Configuring and calculating transaction fees in the emulator.
+-- | Calculating transaction fees in the emulator.
 module Ledger.Fee(
-  FeeConfig (..)
-, calcFees
+  estimateTransactionFee,
+  makeTransactionBodyAutoBalance
 ) where
 
-import Data.Aeson (FromJSON, ToJSON)
-import Data.Default (Default (def))
-import GHC.Generics (Generic)
-import Ledger.Index (minFee)
-import Plutus.V1.Ledger.Ada (Ada)
-import Plutus.V1.Ledger.Ada qualified as Ada
+import Cardano.Api.Shelley qualified as C.Api
+import Cardano.Ledger.BaseTypes (Globals (systemStart))
+import Data.Bifunctor (bimap, first)
+import Data.Map qualified as Map
+import Ledger.Address (Address, PaymentPubKeyHash)
+import Ledger.Params (EmulatorEra, Params (pNetworkId, pProtocolParams), emulatorEraHistory, emulatorGlobals)
+import Ledger.Tx (ToCardanoError (TxBodyError), Tx)
+import Ledger.Tx.CardanoAPI (CardanoBuildTx (..), getCardanoBuildTx, toCardanoAddressInEra, toCardanoTxBodyContent)
+import Ledger.Validation (CardanoLedgerError, UTxO (..), makeTransactionBody)
+import Ledger.Value (Value)
+import Plutus.V1.Ledger.Ada (lovelaceValueOf)
+
+estimateTransactionFee
+  :: Params
+  -> UTxO EmulatorEra
+  -> [PaymentPubKeyHash]
+  -> Tx
+  -> Either CardanoLedgerError Value
+estimateTransactionFee params utxo requiredSigners tx = do
+  txBodyContent <- first Right $ toCardanoTxBodyContent params requiredSigners tx
+  let nkeys = C.Api.estimateTransactionKeyWitnessCount (getCardanoBuildTx txBodyContent)
+  txBody <- makeTransactionBody params utxo txBodyContent
+  case C.Api.evaluateTransactionFee (pProtocolParams params) txBody nkeys 0 of
+    C.Api.Lovelace fee -> pure $ lovelaceValueOf fee
+
 
--- | Datatype to configure the fee in a transaction.
---
--- The fee for a transaction is typically: 'fcConstantFee + 'fcScriptsFeeFactor' *
--- <SIZE_DEPENDANT_SCRIPTS_FEE>.
-data FeeConfig =
-    FeeConfig
-        { fcConstantFee      :: Ada    -- ^ Constant fee per transaction in lovelace
-        , fcScriptsFeeFactor :: Double -- ^ Factor by which to multiply the size-dependent scripts fee
-        }
-    deriving (Show, Eq, Generic, ToJSON, FromJSON)
+makeTransactionBodyAutoBalance
+  :: Params
+  -> UTxO EmulatorEra
+  -> CardanoBuildTx
+  -> Address
+  -> Either CardanoLedgerError (C.Api.Tx C.Api.AlonzoEra)
+makeTransactionBodyAutoBalance params utxo (CardanoBuildTx txBodyContent) pChangeAddr = first Right $ do
```

Comment:
Shouldn't it be `makeTransactionWithAutoBalancedBody`?

Because I read it as "make the transaction body auto balance", but we return the transaction as a result, and make the transaction with auto balanced body.

Just clarifying.

Reply:
Yeah I'm just sticking with the name from cardano-api. I think it is meant to be read as "makeTransactionBody:, the AutoBalance version". I'll add a comment.

Reply:
No actually you are right, I forgot I added a makeSignedTransaction call at the end. So `makeAutoBalancedTransaction` it is!